File: socket/js/api_v4.js
const axios = require('axios');
const crypto = require('crypto');
const jwt = require('jsonwebtoken');

const keyId = "";
const publicKey = "";
const privateKey = "";
const apiUrl = 'https://api.youngplatform.com';

getBalance().then((data) => {
    console.log(data);
    return cancelOrder("1");
})
.then((data) => {
    console.log(data);
})
.catch((e) => {
    console.log(e);
});

async function getBalance() {
    try {
        let url = `${apiUrl}/api/v4/private/balances`;
        let token = generateJwt(publicKey, privateKey, "");
        let headers ={
            "Authorization": token,
            "X-Api-Key-Id": keyId,
        }
        var data = (await axios.get(url, { headers: headers })).data;
        return data;
    } catch (e) {
        console.log(e);
        throw e;
    }
}

async function cancelOrder(orderId) {
    try {
        let url = `${apiUrl}/api/v4/private/cancel`;
        let body = {
            "orderId": orderId
        }
        let bodyString = JSON.stringify(body);
        let token = generateJwt(publicKey, privateKey, bodyString);
        let headers ={
            "Authorization": token,
            "X-Api-Key-Id": keyId,
            "Content-Type": "application/json"
        }
        var data = (await axios.post(url, bodyString, {headers: headers})).data;
        return data;
    }
    catch (e) {
        if (e.response.status == 400) {
            return e.response.data;
        }
        throw e;
    }
}


function generateJwt(publicKey, privateKey, payload) {
    let iat = Math.floor(Date.now() / 1000);
    let exp = iat + 30;
    if (payload == undefined) {
        // if GET request or no payload, use empty string
        payload = "";
    }
    let hashPayload = crypto.createHash('sha256').update(payload).digest('hex');
    const jwtClaims = {
        "sub": publicKey,
        "iat": iat,
        "exp": exp,
        "hash_payload": hashPayload
    };
    let token = jwt.sign(jwtClaims, privateKey, { algorithm: 'HS256' });
   return token;
}

File: socket/python/api_v4.py
import time
import hashlib
import jwt
import requests
import json

key_id = None # Replace with your key ID string
public_key = None # Replace with your public key string
private_key = None # Replace with your private key string

def generate_jwt(public_key, private_key, payload=None):
    # Get the current timestamp and set expiration (30 seconds from now)
    iat = int(time.time())
    exp = iat + 30

    # If payload is None, use an empty string
    if payload is None:
        payload = ""

    # Generate a SHA-256 hash of the payload
    hash_payload = hashlib.sha256(payload.encode('utf-8')).hexdigest()

    # Define the JWT claims
    jwt_claims = {
        "sub": public_key,
        "iat": iat,
        "exp": exp,
        "hash_payload": hash_payload
    }

    # Generate the JWT token
    token = jwt.api_jwt.encode(jwt_claims, private_key, algorithm="HS256")

    return token



# generate a JWT token for GET request
jwtToken = generate_jwt(public_key, private_key)
headers = {
    "Authorization": jwtToken,
    "X-Api-Key-Id": key_id
}

req_url = "https://api.youngplatform.com/api/v4/private/balances"
response = requests.get(req_url, headers=headers)

if response.status_code != 200:
    print(f"Error: {response.status_code}")
    print(response.text)
    exit()
body = response.json()

print(body)

# jwt for POST request

# payload for cancel order
cancel_order_payload = {
    "orderId": 100
}

# generate a JWT token for POST request
body_string = json.dumps(cancel_order_payload)
jwtTokenBody = generate_jwt(public_key, private_key, body_string)

headers = {
    "Authorization": jwtTokenBody,
    "X-Api-Key-Id": key_id,
    "Content-Type": "application/json"
}
req_url = "https://api.youngplatform.com/api/v4/private/cancel"
response = requests.post(req_url, headers=headers, data=body_string)

if response.status_code != 200:
    print(f"Error: {response.status_code}")
    print(response.text)
    exit()